src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSMessagingUtils.cs

```diff
@@ -0,0 +1,141 @@
+// <copyright file="AWSMessagingUtils.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Generic;
+using System.Linq;
+using Amazon.Lambda.SNSEvents;
+using Amazon.Lambda.SQSEvents;
+using Newtonsoft.Json;
+using OpenTelemetry.Context.Propagation;
+
+namespace OpenTelemetry.Instrumentation.AWSLambda.Implementation;
+
+internal class AWSMessagingUtils
+{
+    // SNS attribute types: https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
+    private const string SnsAttributeTypeString = "String";
+    private const string SnsAttributeTypeStringArray = "String.Array";
+    private const string SnsMessageAttributes = "MessageAttributes";
+
+    internal static PropagationContext ExtractParentContext(SQSEvent sqsEvent)
+    {
+        if (sqsEvent == null)
+        {
+            return default;
+        }
+
+        // We assume there can be only one parent that's why we consider only a single (the last) record as the carrier.
+        var message = sqsEvent.Records.LastOrDefault();
+        return ExtractParentContext(message);
+    }
+
+    internal static PropagationContext ExtractParentContext(SQSEvent.SQSMessage sqsMessage)
+    {
+        if (sqsMessage == null)
+        {
+            return default;
+        }
+
+        // SQS subscribed to SNS topic with raw delivery disabled case, i.e. SNS record serialized into SQS body.
+        // https://docs.aws.amazon.com/sns/latest/dg/sns-large-payload-raw-message-delivery.html
+        SNSEvent.SNSMessage snsMessage = GetSnsMessage(sqsMessage);
```

Comment:
I think before checking for the embedded SNS message, we should first try the SQS MessageAttributes, not the other way round.

Reply:
The check order has been swapped.